slidechain/postpegout.go

```diff
@@ -15,32 +15,35 @@ import (
 	"github.com/stellar/go/xdr"
 )
 
-func (c *Custodian) doPostPegOut(ctx context.Context, assetXDR, anchor, txid []byte, amount, seqnum int64, peggedOut pegOutState, exporter, tempAddr string, pubkey []byte) error {
+func (c *Custodian) doPostPegOut(ctx context.Context, p pegOut, txid []byte) error {
+	// The contract needs a non-zero selector to retire funds if the peg-out succeeded.
+	// Else, it requires a zero selector so the funds are returned.
+	var selector int64
+	if p.State == pegOutOK {
+		selector = 1
+	}
+
+	// The refdata stored in the database has pegged-out state equal to 0.
+	// We set the pegOut struct state to this value temporarily.
+	// This lets us correctly reconstruct the contract's stack and call it.
+	currentState := p.State
+	p.State = 0
```

Comment:
This has what Eric Rykwalder used to refer to as "code smell."

A slightly better approach would be to leave `p` alone and mutate a duplicate instead, which means we don't have to remember to put anything back the way it was:

```
p2 := p
p2.State = 0
...
refdata, err := json.Marshal(p2)
```

Better still would be to figure out some way that this hack isn't needed to begin with. Why does `p.State` diverge from what's in the db? Does that mean that `State` should be stored separately from the other items in `p`?

Reply:
I think I was unclear when I explained the issue before, sorry about that.  `State` is already stored separately from the other items in `p`, via the `pegged_out` field in `exports`. `p.State` is consistent with that stored state at the beginning of this function.

However, it differs from the state stored in the JSON encoding of `p`; that is stored in the `pegout_json` field (formerly `ref`)  in the `exports` table.  I think the fix to this is simple and the same as one you suggested elsewhere: instruct the JSON package to skip encoding `State` in the JSON. This will ensure that the JSONs are consistent.